src/compiler/checker.ts

```diff
@@ -12305,6 +12412,17 @@ namespace ts {
             if (node.questionToken && isBindingPattern(node.name) && func.body) {
                 error(node, Diagnostics.A_binding_pattern_parameter_cannot_be_optional_in_an_implementation_signature);
             }
+            if ((<Identifier>node.name).text === "this") {
+                if (indexOf(func.parameters, node) !== 0) {
+                    error(node, Diagnostics.A_this_parameter_must_be_the_first_parameter);
+                }
+                if (func.kind === SyntaxKind.Constructor || func.kind === SyntaxKind.ConstructSignature) {

```

Comment:
Re posting, since the other comment got removed. 

you have `ConstructSignature` covered, i.e. `{ new(this: T): T; }`, i am talking about `ConstructorType` i.e. `var x: new (this: number)=> number.` this still fails on the latest from this branch.